cezo_fl/util/prepare_settings.py

```diff
@@ -0,0 +1,211 @@
+    lr: float,
+    momentum: float,
+    large_model: str | None,
+    lora: bool = False,
+    lora_alpha: int = 16,
+    lora_r: int = 8,
+) -> tuple[AllModel, torch.optim.SGD]:
+    torch_dtype = get_torch_dtype(model_dtype)
+    model: AllModel
+    if dataset == "mnist":
+        model = CNN_MNIST().to(torch_dtype)
+        optimizer = torch.optim.SGD(
+            model_helpers.get_trainable_model_parameters(model),
+            lr=lr,
+            weight_decay=1e-5,
+            momentum=momentum,
+        )
+    elif dataset == "cifar10":
+        model = LeNet().to(torch_dtype)
+        optimizer = torch.optim.SGD(
+            model_helpers.get_trainable_model_parameters(model),
+            lr=lr,
+            weight_decay=5e-4,
+            momentum=momentum,
+        )
+    elif dataset == "fashion":
+        model = CNN_FMNIST().to(torch_dtype)
+        optimizer = torch.optim.SGD(
+            model_helpers.get_trainable_model_parameters(model),
+            lr=lr,
+            weight_decay=1e-5,
+            momentum=momentum,
+        )
+    elif dataset == "shakespeare":
+        model = CharLSTM().to(torch_dtype)
+        optimizer = torch.optim.SGD(
+            model_helpers.get_trainable_model_parameters(model),
+            lr=lr,
+            momentum=0.9,
+            weight_decay=5e-4,
+        )
+    elif dataset in LM_TEMPLATE_MAP.keys():
+        assert large_model in SUPPORTED_LLM
+        hf_model_name = SUPPORTED_LLM[large_model]
+        model = AutoModelForCausalLM.from_pretrained(hf_model_name, torch_dtype=torch_dtype)
+        model.model_name = large_model
+        if lora:
+            # this step initialize lora parameters, which should be under control of seed
+            lora_config = LoraConfig(
+                r=lora_r,
+                lora_alpha=lora_alpha,
+                target_modules=["q_proj", "v_proj"],
+            )
+            model = get_peft_model(model, lora_config).to(torch_dtype)
+
+        if dataset in ["sst2", "cb", "wsc", "wic", "multirc", "rte", "boolq"]:
+            optimizer = torch.optim.SGD(
+                model_helpers.get_trainable_model_parameters(model),
+                lr=lr,
+                momentum=0,
+                weight_decay=5e-4,
+            )
+        elif dataset in ["squad", "drop", "xsum"]:
+            optimizer = torch.optim.SGD(
+                model_helpers.get_trainable_model_parameters(model),
+                lr=lr,
+                momentum=0,
+                weight_decay=0,
+            )
+        else:
+            raise ValueError(f"Dataset {dataset} is not supported")
+    else:
+        raise Exception(f"Dataset {dataset} is not supported")
+
+    return model, optimizer
+
+
+@dataclass
+class ModelInferences:
+    train_inference: Callable
+    test_inference: Callable
+
+
+@dataclass
+class MetricPacks:
+    train_loss: Callable
+    train_acc: Callable
+    test_loss: Callable
+    test_acc: Callable
```

Comment:
looks good! Optional comment, we can let `train_acc` and `test_loss` to be optional and if the downstream code find it is None, just append None to the records?

Reply:
would prefer to keep this a strict structure, but we can make the train_acc and test_loss Callable do nothing when we define them and return Torch.Tensor(0.0)?

Like this

```
MetricPacks(train_loss = cross_entropy, lambda x, y: return torch.Tensor(0.0), test_loss =  lambda x, y: return torch.Tensor(0.0), test_acc = accuracy)
```

Reply:
I see what you mean, you want to let downstream code to skip those steps

Reply:
Yeah, I meant
```python
@dataclass
class MetricPacks:
    train_loss: Callable
    train_acc: Callable | None
    test_loss: Callable | None
    test_acc: Callable
```
But i think `test_loss =  lambda x, y: return torch.Tensor(0.0)` is also good.

Reply:
tried to implement a check for train_acc and test_loss in client and server, but found out we actually need these 2 fields in the tensorboard.
Thus decide to use `test_loss =  lambda x, y: return torch.Tensor(0.0)`